R/avworkspace-deprecated.R

```diff
@@ -58,15 +58,19 @@ NULL
 avworkspaces <-
```

Comment:
Just picking this as an example for discussion.

It seems like it is desirable for the AnVIL package to have a function that can list workspaces. And `AnVIL::avworkspaces()` seems like a good function name.

This function uses the Terra API. Is the Terra API GCP-specific? If not, then this function will work in either Azure or GCP, so does not need to be deprecated. 

If the Terra API is GCP-specific, then instead of deprecating this appropriately named function, I would suggest revising the body of the code along the lines of

```
switch(
    cloud_platform(),
    gcp = AnVILGCP::gcpworkspaces(),
    azure = AnVILAz::azworkspaces(),
    stop("unsupported platform '", cloud_platform(), "'")
)
```

`AnVILGCP::gcpworkspaces()` would contain the current code.

`AnVILAz::azworkspaces()` would contain the azure-equivalent code, or a 'not yet implemented' error.

Hmm, but stepping back a bit more, a different approach would be to think of `avworkspaces()` as a generic, with methods for different cloud platforms. In the current package, one might

```
## base class and interface
platform <- setClass("Platform")
setGeneric(
    "avworkspaces",
     function(platform) standardGeneric("avworkspaces"),
    signature = "platform"   # only dispatch on 'platform'
)
setMethod("avworkspaces", c(platform = "missing"), function(platform) {
    avworkspaces(cloud_platform()) # provide a default value and re-dispatch
})
setMethod("avworkspaces", c(platform = "ANY"), function(platform) {
    stop("'avworkspaces()' is not yet implemented for '", cloud_platform_name(platform), "'")
})

## Platform & implementation
gcp_platform <- setClass("GCPPlatform", contains = "Platform")
## exported constructor
GCPPlatform <- function() {
    gcp_platform()
}
setMethod("avworkspaces", "GCPPlatform", function(platform) {
    ## current implementation
})

az_platform <- setClass("AZPlatform", contains = "Platform")
AZPlatform <- function() {
    az_platform()
}
## when implemented, add...
## setMethod("avworkspaces", "AZPlatform", function(platform) {
##     ## avworkspaces for AZPlatform implementation
## })
```

I know that in the bigger picture there need to be separate packages, so this would lead to

- AnVILBase defining the Platform class, generics, and missing and ANY methods. It would only ever be Import'ed by other packages. It might also contain utility functions shared by AnVILGCP & AnVILAz.
- AnVILGCP and AnVILAz defining methods, plus additional platform-specific functionality
- AnVIL, which would import and then re-export the generics, plus define platform-agnositic functionality.